### DISCUSSION TITLE
Language Support

### DISCUSSION THREAD

**Discussion by Shounak47** [confluentinc/parallel-consumer#754] [Q&A] [answered]
My processing function requires certain libraries that are only available in Python.Is there any way still to use the Parallel Consumer? ( Apart from making a API call from Java->Python in the processing logic)

**Comment by rkolesnev** [comment, marked as answer]
Hi @Shounak47,
IMHO easiest would be to use an API call over Rest or GRPC etc.
There are projects that allow to execute Python code from Java - for example PemJa - https://github.com/alibaba/pemja. 
Flink is using it for PyFlink UDF execution in THREAD mode - some information on that -  https://flink.apache.org/2022/05/06/exploring-the-thread-mode-in-pyflink/ https://nightlies.apache.org/flink/flink-docs-master/docs/dev/python/python_execution_mode/
But it would be quite involved to integrate python code through PemJa.

**Comment by Shounak47** [comment]
Thanks!

**Comment by astubbs** [comment]
@Shounak47 - the answer you accepted is still the one that works today, and there is now a second one on the way.

The maintained fork is building a way for a Python program to use Parallel Consumer without writing the processing in Java: a sidecar your Python process talks to over gRPC, and an in-process path where the engine runs inside your Python interpreter as a GraalVM shared library, with no JVM to install. Your Python-only libraries then run in the processing function where they belong, with PC handling the key ordering, retries and offsets. It is on open pull requests, not in a release - the Python client (astubbs/parallel-consumer#390) and the in-process Python path (astubbs/parallel-consumer#340), tracked by astubbs/parallel-consumer#242, this tracker's #154 - and the "What comes next" section of [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0) lists it among the open pull requests that follow the release. #820 in this tracker is the same question with the same answer.

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
